replica/table.cc

```diff
@@ -437,6 +438,19 @@ void table::enable_off_strategy_trigger() {
 
 future<>
 table::do_add_sstable_and_update_cache(sstables::shared_sstable sst, sstables::offstrategy offstrategy) {
+    if (sst->requires_view_building()) {
+        _sstables_staging.emplace(sst->generation(), sst);
+
+        if (_sstables->contains(sst->generation())) {
+            co_return;
+        }
+
+        // clone staging sstables so their content may be compacted while
+        // views are built.  When done, the hard-linked copy in the staging
+        // subsirectory will be simply unlinked.
+        sst = co_await sst->clone_at(dir());
```

Comment:
Should we reverse this order? Add to `_sstables` first then add to staging set. Will we correctly handle a crash after an sstable is added to staging set only (but not to the regular set)?

Reply:
good point.  In the typical case both the clone at staging and the one in the table base directory need to exist, so this function needs to be exception safe.  It'd be better to cleanup after clone_at failure, since it's easier and cheaper to erase from `_sstables_staging` than to unlink the cloned sstable.

Reply:
Handled in v5